core/src/main/python/model_diff.py

```diff
@@ -0,0 +1,681 @@
+                    last=token.rfind('|')
+                    token=root
+
+
+        # already out of recursive calls, add all entries from current dictionary
+        # resources.JDBCSubsystemResources.* (note it may not have the lower level nodes
+        added_token=token
+        debug('DEBUG: current added token %s' , added_token)
+        if len(added) > 0:
+            for item in added:
+                token=saved_token
+                debug('DEBUG: recursive added token %s item %s ', token, item)
+                all_added.append(token + '|' + item)
+
+        # We don't really care about this, just put something here is enough
+        if len(removed) > 0:
+            for item in removed:
+                token=saved_token
+                debug('DEBUG: removed %s', item)
+                all_removed.append(token + '|' + item)
+        debug('DEBUG: Exiting recursive_changed_detail')
+
+    def is_dict(self,key):
+        if isinstance(self.current_dict[key],dict):
+            return 1
+        else:
+            return 0
+
+    def calculate_changed_model(self):
+        """
+        Calculate the changed model.
+        """
+
+        # This is the top level of changes only
+        #  e.g. from no appDeployments to have appDeployments
+        #       from no resources to have resources
+        #
+        #  changed, added, removed are keys in the dictionary
+        #   i.e. resources, domainInfo, appDeployments, topology
+        #
+        changed = self.changed()
+        added = self.added()
+        removed = self.removed()
+
+        #
+        #  Call recursive for each key
+        #
+        for s in changed:
+            token=s
+            self.recursive_changed_detail(s, token, s)
+            self._add_results(all_changes)
+            self._add_results(all_added)
+            self._add_results(all_removed, True)
+
+        for s in added:
+            token=s
+            self.recursive_changed_detail(s, token, s)
+            self._add_results(all_changes)
+            self._add_results(all_added)
+            # Should not have delete
+
+        # Clean up previous delete first
+        for x in all_removed:
+            all_removed.remove(x)
+
+        # Top level:  e.g. delete all resources, all appDeployments
+
+        for s in removed:
+            token = s
+            self.recursive_changed_detail(s,token, s)
+            self._add_results(all_removed, True)
+
+    def _add_results(self, ar_changes, is_delete=False):
+        """
+        Update the differences in the final model dictionary with the changes
+        :param ar_changes:   Array of changes in delimited format
+        """
+        # The ar_changes is the keys of changes in the piped format
+        #  'resources|JDBCSystemResource|Generic2|JdbcResource|JDBCConnectionPoolParams|TestConnectionsOnReserve
+        #
+
+        # TODO: may be it can be simplified by walking through levels instead of hard coding?
+
```

Comment:
I think if you recurse similar to attributes_test.py, you can check for alias_helper.supports_multiple_mbean_instances(), and that will show you which ones have a name level. with apps and libs, I think everything that has a name level can be deleted.

Reply:
changed to use alias mechanism to walk the tree.